drivers/regulator/regulator_shell.c

```diff
@@ -1,152 +1,404 @@
  */
 
+#include <ctype.h>
 #include <stdlib.h>
-#include <zephyr/kernel.h>
+#include <string.h>
+
+#include <zephyr/device.h>
 #include <zephyr/shell/shell.h>
-#include <zephyr/drivers/uart.h>
 #include <zephyr/drivers/regulator.h>
+#include <zephyr/toolchain.h>
+
+static int strtomicro(char *inp, char units, int32_t *val)
+{
+	size_t len, start, end;
+	int32_t mult, decdiv = 1;
+
+	len = strlen(inp);
+	if (len < 2) {
+		return -EINVAL;
+	}
+
+	/* suffix */
+	if (tolower(inp[len - 1]) != units) {
+		return -EINVAL;
+	}
+
+	if ((len > 2) && (inp[len - 2] == 'u')) {
+		mult = 1;
+		end = len - 3;
+	} else if ((len > 2) && (inp[len - 2] == 'm')) {
+		mult = 1000;
+		end = len - 3;
+	} else if (isdigit(inp[len - 2]) > 0) {
+		mult = 1000000;
+		end = len - 2;
+	} else {
+		return -EINVAL;
+	}
+
+	/* optional prefix (sign) */
+	if (inp[0] == '-') {
+		mult *= -1;
+		start = 1;
+	} else if (inp[0] == '+') {
+		start = 1;
+	} else {
+		start = 0;
+	}
+
+	/* numeric part */
+	*val = 0;
+	for (size_t i = start; i <= end; i++) {
+		if (isdigit(inp[i]) > 0) {
+			*val = *val * 10 / decdiv +
+			       (int32_t)(inp[i] - '0') * mult / decdiv;
+		} else if (inp[i] == '.') {
+			decdiv = 10;
+		} else {
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static void microtoshell(const struct shell *sh, char unit, int32_t val)
+{
+	if (val > 100000) {
+		shell_print(sh, "%d.%03d %c", val / 1000000,
+			    (val % 1000000) / 1000, unit);
+	} else if (val > 1000) {
+		shell_print(sh, "%d.%03d m%c", val / 1000, val % 1000, unit);
+	} else {
+		shell_print(sh, "%d u%c", val, unit);
+	}
+}
+
+static int cmd_enable(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *dev;
+	int ret;
+
+	ARG_UNUSED(argc);
+
+	dev = device_get_binding(argv[1]);
+	if (dev == NULL) {
+		shell_error(sh, "Regulator device %s not available", argv[1]);
+		return -ENODEV;
+	}
+
+	ret = regulator_enable(dev);
+	if (ret < 0) {
+		shell_error(sh, "Could not enable regulator (%d)", ret);
+		return ret;
+	}
 
-LOG_MODULE_REGISTER(regulator_shell, CONFIG_REGULATOR_LOG_LEVEL);
+	return 0;
+}
 
-static int cmd_reg_en(const struct shell *sh, size_t argc, char **argv)
+static int cmd_disable(const struct shell *sh, size_t argc, char **argv)
 {
-	const struct device *reg_dev;
+	const struct device *dev;
 	int ret;
 
-	reg_dev = device_get_binding(argv[1]);
-	if (reg_dev == NULL) {
-		shell_error(sh, "regulator device %s not available", argv[1]);
+	ARG_UNUSED(argc);
+
+	dev = device_get_binding(argv[1]);
+	if (dev == NULL) {
+		shell_error(sh, "Regulator device %s not available", argv[1]);
 		return -ENODEV;
 	}
 
-	ret = regulator_enable(reg_dev);
+	ret = regulator_disable(dev);
 	if (ret < 0) {
-		shell_error(sh, "failed to enable regulator, error %d", ret);
+		shell_error(sh, "Could not disable regulator (%d)", ret);
 		return ret;
 	}
-	shell_print(sh, "enabled regulator");
+
 	return 0;
 }
 
+static int cmd_vlist(const struct shell *sh, size_t argc, char **argv)
```

Comment:
I wanted to get your opinion on an issue I had while using this command with the PCA9420. Since this regulator supports discontinuous voltage ranges (and therefore uses the `liner_range` API), `regulator_count_voltage` will return the number of supported voltage levels (as expected). However, on the PCA9420, this is different from the number of valid voltage indices usable in `regulator_list_voltage`. The result is that not all supported voltages will be listed using this command, since the number of supported voltages is less than the number of indices available for `regulator_list_voltage`

There are a few ways I can think of to resolve this:
- directly record the number of supported voltages for each regulator in the PCA9420 driver, and return this value for an API call to `regulator_count_voltages`
- Modify the `linear_range` API so that `linear_range_group_values_count` counts the number of indices that are valid for a given linear range.

Of the two options, modifying the `linear_range` API makes more sense to me. I'm not sure what cases exist where a user would only want the number of discrete supported values, and not the number of valid indices (since they will probably iterate through the indices if they are using this API)

Reply:
This is actually an oversight that needs to be fixed. I think the fix needs to go into the linear_range API as you suggest. Let me fix it.

Reply:
Fixed the linear_range API. I also adjusted shell so that it doesn't print repeated values. Thanks for reporting this.